cs3api4lab/api/cs3_file_api.py

```diff
@@ -139,67 +114,43 @@ def write_file(self, file_path, content, endpoint=None):
         Write a file using the given userid as access token. The entire content is written
         and any pre-existing file is deleted (or moved to the previous version if supported).
         """
-        #
-        # Prepare endpoint
-        #
-        time_start = time.time()
-        reference = file_utils.get_reference(file_path, endpoint)
-
-        if isinstance(content, str):
-            content_len = len(content)
-        else:
-            content_len = len(content.decode('utf-8'))
-        # providing '0' as size leads to unexpected additional file creation
-        content_size = str(content_len) if content_len > 0 else str(1)
+        file_path = self.lock_manager.resolve_file_path(file_path, endpoint)
 
-        meta_data = types.Opaque(map={"Upload-Length": types.OpaqueEntry(decoder="plain", value=str.encode(content_size))})
-        req = cs3sp.InitiateFileUploadRequest(ref=reference, opaque=meta_data)
-        init_file_upload_res = self.cs3_api.InitiateFileUpload(request=req, metadata=[
-            ('x-access-token', self.auth.authenticate())])
+        time_start = time.time()
 
-        if init_file_upload_res.status.code != cs3code.CODE_OK:
-            self.log.debug('msg="Failed to initiateFileUpload on write" file_path="%s" reason="%s"' % \
-                           (file_path, init_file_upload_res.status.message))
-            raise IOError(init_file_upload_res.status.message)
+        if self.storage_logic.stat(file_path, endpoint) is not None:
+            self.lock_manager.handle_locks(file_path, endpoint)
 
-        self.log.debug(
-            'msg="writefile: InitiateFileUploadRes returned" protocols="%s"' % init_file_upload_res.protocols)
+        content_size = FileUtils.calculate_content_size(content)
+        init_file_upload = self.storage_logic.init_file_upload(file_path, endpoint, content_size)
 
-        #
-        # Upload
-        #
         try:
-            protocol = [p for p in init_file_upload_res.protocols if p.protocol == "simple"][0]
-            headers = {
-                'Tus-Resumable': '1.0.0',
-                'File-Path': file_path,
-                'File-Size': content_size,
-                'x-access-token': self.auth.authenticate(),
-                'X-Reva-Transfer': protocol.token
-            }
-            put_res = requests.put(url=protocol.upload_endpoint, data=content, headers=headers)
+            upload_reponse = self.storage_logic.upload_content(file_path, content, content_size, init_file_upload)
 
+            self.lock_manager.handle_locks(file_path, endpoint)
```

Comment:
Here I decided to lock again, because uploading a file might take some time.

Reply:
Why don't we just lock at the end? Even if it takes some time, it shouldn't be long enough that the locks gets invalid.

Reply:
It's not a good idea in my opinion. It could lead to a case where 2 people start uploading the same file. If anything it should be done in the beginning. But I think it's worth it to lock 2 times, even if we add some extra requests in the process. If you are really concerned about that, I could add some kind of mechanism that would ensure additional locking requests are not made if the previous one was done less than something like 30s/1min ago.

Reply:
Fine, let's see how it behaves